### PR TITLE
feat: sign typed data implementation

### PR DESCRIPTION
**Pull request by mersinvald**

### CODE REVIEW

File: src/util/typed.rs [hunk 1]
```diff
@@ -0,0 +1,2344 @@
...
+        #[test]
+        fn test_1() {
+            do_test_hash(
+                r#"{"types":{"EIP712Domain":[{"name":"name","type":"string"},{"name":"version","type":"string"},{"name":"chainId","type":"uint256"},{"name":"verifyingContract","type":"address"}],"Person":[{"name":"name","type":"string"},{"name":"wallet","type":"address"}],"Mail":[{"name":"from","type":"Person"},{"name":"to","type":"Person"},{"name":"contents","type":"string"}]},"primaryType":"Mail","domain":{"name":"Ether Mail","version":"1","chainId":1,"verifyingContract":"0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"},"message":{"from":{"name":"Cow","wallet":"0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"},"to":{"name":"Bob","wallet":"0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"},"contents":"Hello, Bob!"}}"#,
+                r#"vmCa7jQ/s8Syjh355jL8pk/Prt4g8C6GJE793zCVe9I="#,
+            );
+        }
+        #[test]
+        fn test_2() {
+            do_test_hash(
```
Comment (r8d8) [inline src/util/typed.rs]: Are there any difference between test 2 and 3, 4 and 5 ?

File: src/util/typed.rs [hunk 2]
```diff
@@ -0,0 +1,2344 @@
...
+        fn do_test_hash(typed_data: &str, expected: &str) {
+            let input =
+                serde_json::from_str(typed_data).expect("failed to deserialize 'typed_data' json");
+            let got = hash(input).unwrap();
+            let got_base64 = base64::encode(&got);
+            assert_eq!(got_base64, expected);
+        }
+
+        #[test]
+        fn test_1() {
```
Comment (r8d8) [inline src/util/typed.rs:337]: Better to use meaningfull test name - description what these test checks.

File: src/util/typed.rs (hunk 1 above)
Comment (mersinvald) [reply src/util/typed.rs]: Forgot the dedup stage, yeah, will do.

File: src/util/typed.rs (hunk 2 above)
Comment (mersinvald) [reply src/util/typed.rs:337]: Tests are literally a dump of input/output from functions in eth-sig js library

File: src/util/typed.rs (hunk 1 above)
Comment (mersinvald) [reply src/util/typed.rs]: Deduplicated it, diff got a lot smaller :)

**Pull request merged**